File: app/orchestrator.py
"""Orchestrator: prepara o contexto e delega ao agente flexível.

Com o modelo de skills, o roteamento por classificador de palavra-chave deixou de
existir — quem decide QUAL capacidade usar é o LLM, chamando as skills do agente
(ver app.agents.skilled). O Orchestrator ficou fino e cuida só do transversal:

  1. Trunca o histórico para as últimas HISTORY_LIMIT mensagens (economia de tokens).
  2. Conversa longa demais (acima de max_turns) -> fallback estático (handoff).
  3. Senão, executa o SkilledAgent.
  4. Anexa metadados de roteamento (intent derivado das skills usadas, tokens...).
  5. Qualquer exceção vira um handoff gracioso (o str(e) fica só no log).
"""

import logging
from typing import TYPE_CHECKING, List

from app.agents import FallbackAgent, SkilledAgent
from app.config import HISTORY_LIMIT
from app.domain import AgentConfig, ChatMessage
from app.messages import ERROR_INTERNAL

if TYPE_CHECKING:
    from app.llm import LLMClient

log = logging.getLogger("blip-agent.orchestrator")

# Skills de catálogo — usado só para derivar um "intent" legível dos metadados.
_CATALOG_SKILLS = {"check_stock", "search_products", "list_products",
                   "reserve_stock", "check_catalog"}


def _intent_from_result(result: dict) -> str:
    """Rótulo legível da capacidade usada (para o painel/avaliação). Substitui o
    intent do antigo classificador — agora é derivado das skills efetivamente
    chamadas, não de um palpite por palavra-chave."""
    tools = set(result.get("tools_called") or [])
    source = result.get("source")
    if source in ("faq_shortcut", "llm_rag") or "knowledge_search" in tools:
        return "faq"
    if source == "support_escalation" or "escalate_to_human" in tools:
        return "support"
    if tools & _CATALOG_SKILLS:
        return "order"
    return "chat"


class Orchestrator:
    def __init__(self, agent_config: AgentConfig, llm: "LLMClient"):
        self.config = agent_config
        self.agent = SkilledAgent(agent_config, llm)
        self.fallback = FallbackAgent(agent_config, llm)

    async def process(self, message: str, history: List[ChatMessage]) -> dict:
        try:
            recent_history = history[-HISTORY_LIMIT:]

            # O comprimento TOTAL da conversa decide o limite de turnos, mesmo que
            # só as últimas mensagens vão ao LLM.
            if len(history) > self.config.max_turns:
                agent_result = await self.fallback.execute(message, recent_history)
                agent_used = "fallback"
            else:
                agent_result = await self.agent.execute(message, recent_history)
                agent_used = "skilled"

            result = agent_result.to_dict()
            # Sem classificador: o "intent" vem das skills usadas e a confiança é
            # 1.0 (a decisão é do LLM, não mais um score de palavra-chave).
            result.update(intent=_intent_from_result(result), confidence=1.0,
                          agent_used=agent_used)

            log.info("agent=%s intent=%s handoff=%s tokens=%d tools=%s",
                     agent_used, result["intent"], result["should_handoff"],
                     result["tokens_used"], result.get("tools_called"))
            return result

        except Exception as e:  # noqa: BLE001 — qualquer falha degrada em handoff.
            # O detalhe técnico fica só no log (correlacionado por request/tenant);
            # a resposta ao usuário usa a mensagem genérica e NÃO vaza str(e).
            log.exception("Erro no orchestrator: %s", e)
            return {
                "response": ERROR_INTERNAL,
                "should_handoff": True,
                "handoff_reason": "Erro interno no processamento.",
                "source": "error",
                "intent": "error",
                "agent_used": "fallback",
                "confidence": 0.0,
                "tokens_used": 0,
            }

File: src/server/orchestrator.ts
/**
 * Orchestrator (porta `app/orchestrator.py`): prepara o contexto e delega ao
 * agente flexível. Fino — só o transversal:
 *   1. Trunca o histórico (HISTORY_LIMIT).
 *   2. Conversa longa demais (> maxTurns) -> FallbackAgent (handoff estático).
 *   3. Senão, executa o SkilledAgent.
 *   4. Anexa metadados (intent derivado das skills usadas, tokens...).
 *   5. Qualquer exceção vira handoff gracioso (o detalhe fica só no log).
 *
 * Sem `lru_cache`: o Orchestrator é instanciado por request (barato); o LLMClient
 * é o singleton de getLlm().
 */

import { FallbackAgent, SkilledAgent } from "./agents";
import { HISTORY_LIMIT, PROMPT_GUARD_MODEL } from "./config";
import type { AgentConfig, AgentResult, ChatMessage } from "./domain";
import type { LLMClient } from "./llm";
import { getLogger } from "./logging";
import { ERROR_INTERNAL, injectionRefusal } from "./messages";
import { detectInjectionAcross } from "./security/injection";
import { sanitizeUntrusted } from "./security/sanitize";

const log = getLogger("blip-agent.orchestrator");

// Skills de catálogo — só para derivar um "intent" legível dos metadados.
const CATALOG_SKILLS = new Set(["check_stock", "search_products", "list_products", "reserve_stock", "check_catalog"]);

export interface ProcessResult {
  response: string;
  shouldHandoff: boolean;
  handoffReason: string | null;
  source: string;
  tokensUsed: number;
  toolsCalled: string[];
  ragChunksUsed: number;
  ragSources: string[];
  intent: string;
  agentUsed: string;
  confidence: number;
}

function intentFromResult(result: AgentResult): string {
  const tools = new Set(result.toolsCalled);
  const source = result.source;
  if (source === "faq_shortcut" || source === "llm_rag" || tools.has("knowledge_search")) return "faq";
  if (source === "support_escalation" || tools.has("escalate_to_human") || tools.has("create_ticket")) return "support";
  for (const s of CATALOG_SKILLS) {
    if (tools.has(s)) return "order";
  }
  return "chat";
}

export class Orchestrator {
  private readonly config: AgentConfig;
  private readonly llm: LLMClient;
  private readonly agent: SkilledAgent;
  private readonly fallback: FallbackAgent;

  constructor(agentConfig: AgentConfig, llm: LLMClient) {
    this.config = agentConfig;
    this.llm = llm;
    this.agent = new SkilledAgent(agentConfig, llm);
    this.fallback = new FallbackAgent(agentConfig, llm);
  }

  async process(message: string, history: ChatMessage[]): Promise<ProcessResult> {
    try {
      const recentHistory = history.slice(-HISTORY_LIMIT);

      // 1. Detecção de contaminação no texto BRUTO (antes de sanitizar — senão os
      //    sinais, como invisíveis e tokens de template, já teriam sumido).
      const verdict = detectInjectionAcross([message, ...recentHistory.map((m) => m.content)]);
      if (verdict.suspicious) {
        log.warn(`prompt suspeito: score=${verdict.score} reasons=${JSON.stringify(verdict.reasons)}`);
      }

      // 2. Classificador opcional em cascata: só quando o heurístico já suspeita
      //    E há modelo configurado. Fail-open (erro -> "unknown" -> sem efeito).
      let hardened = verdict.flagged;
      let refuse = verdict.refuse;
      if (!refuse && verdict.suspicious && PROMPT_GUARD_MODEL) {
        const guard = await this.llm.classifyInjection(message);
        if (guard === "injection") {
          hardened = true;
          refuse = true; // o modelo confirmou manipulação que a heurística não fechou
          log.warn("Guard de injeção classificou a entrada como maliciosa.");
        }
      }

      // 3. Manipulação de ALTA confiança: responde NO PAPEL, sem chamar o LLM
      //    (determinístico -> impossível de sofrer jailbreak). Não faz handoff:
      //    o bot permanece como o assistente configurado e convida a um pedido real.
      if (refuse) {
        log.warn(`prompt injection recusado (determinístico): reasons=${JSON.stringify(verdict.reasons)}`);
        return {
          response: injectionRefusal(this.config.name),
          shouldHandoff: false,
          handoffReason: null,
          source: "injection_refusal",
          tokensUsed: 0,
          toolsCalled: [],
          ragChunksUsed: 0,
          ragSources: [],
          intent: "chat",
          agentUsed: "guard",
          confidence: 1.0,
        };
      }

      // 4. Sanitiza tudo que efetivamente vai ao LLM (mensagem + histórico).
      const cleanMessage = sanitizeUntrusted(message);
      const cleanHistory = recentHistory.map((m) => ({ ...m, content: sanitizeUntrusted(m.content) }));

      // O comprimento TOTAL da conversa decide o limite de turnos.
      let result: AgentResult;
      let agentUsed: string;
      if (history.length > this.config.maxTurns) {
        result = await this.fallback.execute(cleanMessage, cleanHistory, { hardened });
        agentUsed = "fallback";
      } else {
        result = await this.agent.execute(cleanMessage, cleanHistory, { hardened });
        agentUsed = "skilled";
      }

      const intent = intentFromResult(result);
      log.info(
        `agent=${agentUsed} intent=${intent} handoff=${result.shouldHandoff} ` +
          `tokens=${result.tokensUsed} tools=${JSON.stringify(result.toolsCalled)}`,
      );
      return { ...result, intent, confidence: 1.0, agentUsed };
    } catch (e) {
      // O detalhe técnico fica só no log; a resposta usa a mensagem genérica.
      log.exception("Erro no orchestrator:", e);
      return {
        response: ERROR_INTERNAL,
        shouldHandoff: true,
        handoffReason: "Erro interno no processamento.",
        source: "error",
        tokensUsed: 0,
        toolsCalled: [],
        ragChunksUsed: 0,
        ragSources: [],
        intent: "error",
        agentUsed: "fallback",
        confidence: 0.0,
      };
    }
  }
}
